src/api/app/controllers/concerns/webui/notifications_filter.rb

```diff
@@ -1,33 +1,56 @@
-    when 'incoming_requests'
-      notifications.for_incoming_requests(User.session!)
-    when 'outgoing_requests'
-      notifications.for_outgoing_requests(User.session!)
-    when 'relationships_created'
-      notifications.for_relationships_created
-    when 'relationships_deleted'
-      notifications.for_relationships_deleted
-    when 'build_failures'
-      notifications.for_failed_builds
-    when 'reports'
-      notifications.for_reports
-    when 'workflow_runs'
-      notifications.for_workflow_runs
-    when 'appealed_decisions'
-      notifications.for_appealed_decisions
-    else # all
-      notifications
-    end
+    return notifications if filter_type == ['all']
+
+    relations_type = []
+    relations_type << notifications.for_comments if filter_type.include?('comments')
+    relations_type << notifications.for_requests if filter_type.include?('requests')
+    relations_type << notifications.for_incoming_requests(User.session) if filter_type.include?('incoming_requests')
+    relations_type << notifications.for_outgoing_requests(User.session) if filter_type.include?('outgoing_requests')
+    relations_type << notifications.for_relationships_created if filter_type.include?('relationships_created')
+    relations_type << notifications.for_relationships_deleted if filter_type.include?('relationships_deleted')
+    relations_type << notifications.for_build_failures if filter_type.include?('build_failures')
+    relations_type << notifications.for_reports if filter_type.include?('reports')
+    relations_type << notifications.for_workflow_runs if filter_type.include?('workflow_runs')
+    relations_type << notifications.for_appealed_decisions if filter_type.include?('appealed_decisions')
+
+    notifications = notifications.merge(relations_type.inject(:or)) unless relations_type.empty?
+    notifications
   end
   # rubocop:enable Metrics/CyclomaticComplexity
+  # rubocop:enable Metrics/PerceivedComplexity
+
+  def filter_notifications_by_state(notifications, filter_state)
+    if filter_state.include?('unread') && filter_state.include?('read')
+      notifications.merge(notifications.unread.or(notifications.read))
+    elsif filter_state.include?('read')
+      notifications = notifications.read
+    else
+      notifications = notifications.unread
+    end
+    notifications
+  end
+
+  def filter_notifications_by_project(notifications, filter_project)
+    relations_project = filter_project.map do |project_name|
+      notifications.for_project_name(project_name)
+    end
+
+    notifications = notifications.merge(relations_project.inject(:or)) unless relations_project.empty?
+    notifications
+  end
+
+  def filter_notifications_by_group(notifications, filter_group)
+    relations_group = filter_group.map do |group_title|
+      notifications.for_group_title(group_title)
+    end
+
+    notifications = notifications.merge(relations_group.inject(:or)) unless relations_group.empty?
+    notifications
+  end
```

Comment:
```suggestion
  def filter_notifications_by_group(notifications, filter_group)
    relations_group = filter_group.map do |group_title|
      notifications.for_group_title(group_title)
    end

    notifications.merge(relations_group.inject(:or)) unless relations_group.empty?
  end
```
Same here, you don't need to overwrite the incoming `notifications` parameter if you set the result of this function to `@notifications` down below...

Reply:
But if `relations_group.empty?` is `true` the line is skipped and the function returns nothing, and that crashes in the following code-lines.

Reply:
It could be written like

`relations_group.empty? ? notifications : notifications.merge(relations_group.inject(:or))`

maybe it is more clear?

Reply:
Hm, you are right again, forget about it, I don't really like ternaries